[CUDA12] Clean up deprecated APIs

xw285cornell opened the pull request:
See #91122
Summary:
Some APIs are deprecated in newer version of CUDA.
* cudaGraphInstantiate:
From:
```
cudaGraphInstantiate ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, cudaGraphNode_t* pErrorNode, char* pLogBuffer, size_t bufferSize )
```
To
```
__host__​cudaError_t cudaGraphInstantiate ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, unsigned long long flags = 0 )
```
* cudaProfilerInitialize: deprecated in cuda 11 and removed in cuda 12

Test Plan: GH CI

Differential Revision: D41469051

xw285cornell commented on aten/src/ATen/cuda/CUDASparseDescriptors.h:
```diff
@@ -143,9 +143,9 @@ class TORCH_CUDA_CPP_API CuSparseDnVecDescriptor
+//AT_USE_HIPSPARSE_GENERIC_52_API() || (AT_USE_CUSPARSE_GENERIC_API() && AT_USE_CUSPARSE_NON_CONST_DESCRIPTORS())
 
-#if AT_USE_CUSPARSE_CONST_DESCRIPTORS()
+#elif AT_USE_CUSPARSE_CONST_DESCRIPTORS()
```
This one is pretty much cosmetic (because I was working on the endif stuff similar to #90897). But I'm curious if we will end up in double definition since it's two independent if. So just add it here

xw285cornell replied:
cc. @xwang233 @IvanYashchuk

ngimel commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -159,10 +159,14 @@ void CUDAGraph::capture_end() {
   AT_CUDA_CHECK(cudaDriverGetVersion(&version));
   if (version < 11040) {
 #endif
+#if CUDA_VERSION >= 12000
```
this would conflict with #91118? I'm fine with landing whatever's fastest to unblock, but let's please coordinate in #91122

xw285cornell replied:
ah ok I wasn't aware of https://github.com/pytorch/pytorch/pull/91118. In this case let's land that one

xw285cornell committed:
[CUDA12] Clean up deprecated APIs (#91050)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/91050

Some APIs are deprecated in newer version of CUDA.
* cudaGraphInstantiate:
From:
```
cudaGraphInstantiate ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, cudaGraphNode_t* pErrorNode, char* pLogBuffer, size_t bufferSize )
```
To
```
__host__​cudaError_t cudaGraphInstantiate ( cudaGraphExec_t* pGraphExec, cudaGraph_t graph, unsigned long long flags = 0 )
```
* cudaProfilerInitialize: deprecated in cuda 11 and removed in cuda 12

Test Plan: GH CI

Reviewed By: jianyuh

Differential Revision: D41469051

fbshipit-source-id: b4e856717fc48672e7af5d2baba7931a31273c6d
